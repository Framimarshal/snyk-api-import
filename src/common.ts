export const IMPORT_LOG_NAME = 'imported-targets.log';
export const FAILED_LOG_NAME = 'failed-imports.log';
export const FAILED_PROJECTS_LOG_NAME = 'failed-projects.log';
export const FAILED_POLLS_LOG_NAME = 'failed-polls.log';
export const IMPORT_JOBS_LOG_NAME = 'import-jobs.log';
export const IMPORTED_PROJECTS_LOG_NAME = 'imported-projects.log';
export const IMPORTED_BATCHES_LOG_NAME = 'imported-batches.log';
export const IMPORT_JOB_RESULTS = 'import-job-results.log';
export const CREATED_ORG_LOG_NAME = 'created-orgs.log'
export const FAILED_ORG_LOG_NAME = 'failed-to-create-orgs.log'
export const targetProps = [
  'name',
  'appId',
  'functionId',
  'slugId',
  'projectKey',
  'repoSlug',
  // 'id', skip Gitlab ID so we can match against Snyk project data where ID is never returned by Snyk APIs
  'owner',
  'branch',
];
export const targetPropsWithId = [...targetProps, 'id'];

export const UPDATED_BATCHES_LOG_NAME = 'updated-batches.log';