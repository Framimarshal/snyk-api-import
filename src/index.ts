#!/usr/bin/env node

import * as debugLib from 'debug';
const debug = debugLib('snyk:import-projects-script');

export * from './lib';
import { ImportProjects } from './scripts/import-projects';
import { getImportProjectsFile } from './lib/get-import-path';

try {
  const importFile = getImportProjectsFile();
  ImportProjects(importFile);
} catch (e) {
  debug('Failed to kick off import.\n' + e);
  console.error('ERROR! Try running with `DEBUG=snyk* snyk-import`')
}
