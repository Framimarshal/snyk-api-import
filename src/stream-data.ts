import * as debugLib from 'debug';

import * as fs from 'fs';
import split = require('split');

const debug = debugLib('streamData:load-file');
const debugSnyk = debugLib('snyk:load-file');

export async function streamData<DataType>(
  file: string,
  jsonKey: string,
): Promise<DataType[]> {
  let res;
  debug('Trying streamMinifiedJson');
  res = await streamMinifiedJson<DataType>(file, jsonKey);
  if (!res) {
    debugSnyk(
      `Warning: Failed to load as minified JSON, trying to load as beautified JSON with spaces`,
    );
    res = await streamBeautifiedJson<DataType>(file, jsonKey);
  }
  return res;
}

export async function streamMinifiedJson<DataType>(
  file: string,
  arrayKey: string,
): Promise<DataType[]> {
  return new Promise((resolve, reject) => {
    let data: DataType[];
    fs.createReadStream(file)
      .pipe(split())
      .on('data', (lineObj) => {
        if (!lineObj) {
          return;
        }
        try {
          const json = JSON.parse(lineObj);
          data = json[arrayKey];
        } catch (e) {
          debug(`ERROR: Could not find "${arrayKey}" key in json. Make sure the JSON is valid and the key is present`)
          debug(e.message);
        }
      })
      .on('error', (err) => {
        debug('Failed to createReadStream for file: ' + err);
        return reject(err);
      })
      .on('end', async () => resolve(data));
  });
}

export async function streamBeautifiedJson<DataType>(
  file: string,
  arrayKey: string,
): Promise<DataType[]> {
  return new Promise((resolve, reject) => {
    let data: DataType[];
    fs.createReadStream(file)
      .on('data', (jsonData) => {
        if (!jsonData) {
          return;
        }
        try {
          const json = JSON.parse(jsonData);
          data = json[arrayKey];
        } catch (e) {
          debug('ERROR parsing: ', e);
        }
      })
      .on('error', (err) => {
        debug('Failed to createReadStream for file: ' + err);
        return reject(err);
      })
      .on('end', async () => resolve(data));
  });
}
