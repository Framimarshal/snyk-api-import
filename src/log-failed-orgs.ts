import * as debugLib from 'debug';
import * as bunyan from 'bunyan';

import { getLoggingPath } from './lib/get-logging-path';
import { FAILED_ORG_LOG_NAME } from './common';

const debug = debugLib('snyk:create-orgs-script');

export function logFailedOrgs(
  orgs: {
    groupId: string;
    name: string;
    errorMessage: string;
  }[],
  loggingPath = getLoggingPath(),
): void {
  const log = bunyan.createLogger({
    name: 'snyk:create-orgs-script',
    level: 'error',
    streams: [
      {
        level: 'error',
        path: `${loggingPath}/${FAILED_ORG_LOG_NAME}`,
      },
    ],
  });

  orgs.forEach((org) => {
    try {
      const { name, groupId, errorMessage } = org;
      log.error({ name, groupId, errorMessage }, 'Failed to create org');
    } catch (e) {
      debug('Failed to log failed orgs at location: ', e);
      // do nothing
    }
  });
}
