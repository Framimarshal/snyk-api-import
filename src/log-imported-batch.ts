import * as fs from 'fs';
import { getLoggingPath } from './lib/get-logging-path';
import { IMPORTED_BATCHES_LOG_NAME } from './common';

export async function logImportedBatch(
  message: string,
  loggingPath: string = getLoggingPath(),
): Promise<void> {
  try {
    fs.appendFileSync(`${loggingPath}/${IMPORTED_BATCHES_LOG_NAME}`, `${message}\n`);
  } catch (e) {
    // do nothing
  }
}
