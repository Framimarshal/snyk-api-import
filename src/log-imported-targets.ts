import * as fs from 'fs';
import { Target } from './lib/types';
import { getLoggingPath } from './lib/get-logging-path';
import { IMPORT_LOG_NAME } from './common';

export async function logImportedTarget(
  orgId: string,
  integrationId: string,
  target: Target,
  locationUrl: string,
  loggingPath: string = getLoggingPath(),
): Promise<void> {
  try {
    const log = `${integrationId}:${Object.values(target).join(':')}:${locationUrl},`;
    fs.appendFileSync(`${loggingPath}/${orgId}.${IMPORT_LOG_NAME}`, log);
  } catch (e) {
    // do nothing
  }
}
