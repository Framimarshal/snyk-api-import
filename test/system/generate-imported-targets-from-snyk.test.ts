import { exec } from 'child_process';
import * as path from 'path';
import { sep } from 'path';
import { IMPORT_LOG_NAME } from '../../src/common';
import { deleteFiles } from '../delete-files';
const main = './dist/index.js'.replace(/\//g, sep);

describe('`snyk-api-import list:imported <...>`', () => {
  const OLD_ENV = process.env;
  process.env.GITHUB_TOKEN = process.env.GH_TOKEN;

  afterAll(async () => {
    process.env = { ...OLD_ENV };
  });
  it('Shows help text as expected', async (done) => {
    return exec(`node ${main} list:imported help`, (err, stdout) => {
      if (err) {
        throw err;
      }
      expect(err).toBeNull();
      expect(stdout.trim()).toMatchSnapshot();
      done();
    });
  });

  it('Generates Snyk imported targets data as expected for github', async (done) => {
    return exec(
      `node ${main} list:imported --integrationType=github --groupId=${process.env.TEST_GROUP_ID}`,
      {
        env: {
          PATH: process.env.PATH,
          SNYK_TOKEN: process.env.SNYK_TOKEN_TEST,
          SNYK_API: process.env.SNYK_API_TEST,
          SNYK_LOG_PATH: __dirname,
        },
      },
      (err, stdout) => {
        if (err) {
          throw err;
        }
        expect(err).toBeNull();
        expect(stdout.trim()).toMatch(
          `repo(s). Written the data to file: ${path.resolve(
            __dirname,
            'imported-targets.log',
          )}`,
        );
        deleteFiles([path.resolve(__dirname, IMPORT_LOG_NAME)]);
        done();
      },
    );
  }, 10000);
  it('Generates Snyk imported targets data as expected', async (done) => {
    return exec(
      `node ${main} list:imported --integrationType=github --integrationType=github-enterprise --groupId=${process.env.TEST_GROUP_ID}`,
      {
        env: {
          PATH: process.env.PATH,
          SNYK_TOKEN: process.env.SNYK_TOKEN_TEST,
          SNYK_API: process.env.SNYK_API_TEST,
          SNYK_LOG_PATH: __dirname,
        },
      },
      (err, stdout) => {
        if (err) {
          throw err;
        }
        expect(err).toBeNull();
        expect(stdout.trim()).toMatch(
          `target(s). Written the data to file: ${path.resolve(
            __dirname,
            'imported-targets.log',
          )}`,
        );
        deleteFiles([path.resolve(__dirname, IMPORT_LOG_NAME)]);
        done();
      },
    );
  }, 10000);
  it('Shows error when missing groupId', async (done) => {
    return exec(
      `node ${main} list:imported --integrationType=github`,
      (err, stdout) => {
        expect(err).toMatchSnapshot();
        expect(stdout).toEqual('');
        done();
      },
    );
  });
});
