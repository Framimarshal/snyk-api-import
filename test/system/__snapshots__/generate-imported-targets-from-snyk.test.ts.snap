// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`snyk-api-import list:imported <...>\` Shows error when missing groupId 1`] = `
[Error: Command failed: node ./dist/index.js list:imported --integrationType=github
index.js list:imported

List all targets imported in Snyk for a given group & source type. An analysis
is performed on all current orgs and their projects to generate this. The
generated file can be used to skip previously imported targets when running the
\`import\` command

Options:
  --version          Show version number                               [boolean]
  --help             Show help                                         [boolean]
  --groupId          Public id of the group in Snyk (available on group
                     settings)                                        [required]
  --integrationType  The configured integration type (source of the projects in
                     Snyk e.g. Github, Github Enterprise.). This will be used to
                     pick the correct integrationID from each org in Snyk E.g.
                     --integrationType=github
                     --integrationType=github-enterprise
          [required] [choices: "github", "github-enterprise", "bitbucket-cloud"]
                                                                   [default: []]

Missing required argument: groupId
]
`;

exports[`\`snyk-api-import list:imported <...>\` Shows help text as expected 1`] = `
"index.js list:imported

List all targets imported in Snyk for a given group & source type. An analysis
is performed on all current orgs and their projects to generate this. The
generated file can be used to skip previously imported targets when running the
\`import\` command

Options:
  --version          Show version number                               [boolean]
  --help             Show help                                         [boolean]
  --groupId          Public id of the group in Snyk (available on group
                     settings)                                        [required]
  --integrationType  The configured integration type (source of the projects in
                     Snyk e.g. Github, Github Enterprise.). This will be used to
                     pick the correct integrationID from each org in Snyk E.g.
                     --integrationType=github
                     --integrationType=github-enterprise
          [required] [choices: \\"github\\", \\"github-enterprise\\", \\"bitbucket-cloud\\"]
                                                                   [default: []]"
`;
